### ISSUE TITLE
UnneededDisable: Weird false positives

### ISSUE THREAD

**Issue by zor-el** [rubocop-hq/rubocop#6811]
Observed behavior:

``` bash
$ rubocop .
Inspecting 213 files
......W....^C
Exiting... Interrupt again to exit immediately.
.

Offenses:

app/models/account.rb:1:1: W: Lint/UnneededDisable: Unnecessary disabling of Rails/ApplicationRecord.
# rubocop:disable Rails/ApplicationRecord
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
app/models/account.rb:15:17: W: Lint/UnneededDisable: Unnecessary disabling of Rails/Delegate.
  def licensed? # rubocop:disable Rails/Delegate
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

12 files inspected, 2 offenses detected
$ rubocop app/models/account.rb 
Inspecting 1 file
.

1 file inspected, no offenses detected
$ 
```

- `rubocop .` reports false positives  (the cops are NOT disabled in .rubocop.yml)
- `rubocop OFFENDING-FILE.rb` says everything is ok (as it should be).
- Note the `^C` interrupt is just to avoid clutter here. The behavior is the same when rubocop runs uninterrupted.

Expected behavior: `rubocop .` should report no issues, or at least both invocations should report the same offenses for OFFENDING-FILE.rb.

Behavior detected with with rubocop 0.49.1, and reproduced with rubocop 0.64.0

**UPDATE 1:** `rubocop */` shows no offenses either.

**UPDATE 2:** I had to disable caching to reproducibly observe the effect.

**Comment by zor-el** [comment]
Hey @rrosenblum, thanks for responding.

I don't think it's working normally though:

1. AFAIK `rubocop:disable` does not work across files, and `rubocop:disable Rails/ApplicationRecord` is on the 1st line of the file (as indicated by _app/models/account.rb:**1:1**_).

2. Regardless, **shouldn't `rubocop app/models/account.rb` report the same offenses for that file as `rubocop .` ?**

3. I also tried `rubocop app/ spec/ ... [all not excluded folders]` and still no offenses. Only `rubocop .` reports them.

That doesn't seem to add up, or am I missing something very fundamental?

**Comment by jonas054** [comment]
@zor-el I have not been able to reproduce. Could you show us a minimal example of the contents of `app/models/account.rb` plus `.rubocop.yml` that causes the problem to occur?

**Comment by zor-el** [comment]
@jonas054, I can, but note I tried myself to create a minimal case (just .rubocop.yml and account.rb), and could not reproduce the issue either. There is nothing special in account.rb either.

FYI, there are extending config files in the project beyond the main one, but none that affect the `app/models` folder.

``` sh
$ find . | grep 'rubocop.yml'
./tools/.rubocop.yml
./app/models/concerns/.rubocop.yml
./features/.rubocop.yml
./spec/.rubocop.yml
./.rubocop.yml
```

#### account.rb

``` ruby
class Account < ActiveRecord::Base # rubocop:disable Rails/ApplicationRecord

  has_one :license, inverse_of: :account, dependent: :destroy
  has_many :users, inverse_of: :account, dependent: :destroy

  validates :title, presence: true, uniqueness: true
  validates :time_zone, presence: true

  auto_strip_attributes :title, squish: true

  EXPIRATION_TIME = 7.days

  def licensed? # rubocop:disable Rails/Delegate
    license&.licensed?
  end
end
```
#### Main .rubocop.yml

```yaml
AllCops:

  TargetRubyVersion: 2.3
  TargetRailsVersion: 5.0
  UseCache: no

  Exclude:
    - '**/templates/**/*'
    - '**/vendor/**/*'
    - 'Gemfile'
    - 'Guardfile'
    - 'tmp/**/*'
    - 'config/**/*'
    - 'db/**/*'
    - 'spec/*_helper.rb'
    - 'lib/tasks/**/*'
    - 'try/**/*'
    - 'other/**/*'
    - '**/node_modules/**/*'
    - 'bin/*'

  DisplayCopNames: yes
  DisplayStyleGuide: yes
  ExtraDetails: yes

Rails:
  Enabled: yes

Style/TernaryParentheses:
  Enabled: no

Style/FrozenStringLiteralComment:
  Enabled: no

Style/MutableConstant:
  Enabled: no

Style/Not:
  Enabled: no

Style/StringLiterals:
  Enabled: no

Style/Documentation:
  Enabled: no

Style/SignalException:
  Enabled: no

Style/For:
  Enabled: no

Style/AndOr:
  Enabled: no

Style/FormatString:
  EnforcedStyle: percent

Style/TrailingCommaInArguments:
  Enabled: no

Style/TrailingCommaInLiteral:
  Enabled: no

Style/EmptyMethod:
  Enabled: no

Style/SymbolArray:
  Enabled: no

Style/PercentLiteralDelimiters:
  Enabled: no

Style/LambdaCall:
  Enabled: no

Layout/SpaceBeforeBlockBraces:
  Enabled: no

Layout/EmptyLinesAroundClassBody:
  Enabled: no

Layout/EmptyLinesAroundModuleBody:
  Enabled: no

Layout/EmptyLinesAroundBlockBody:
  Enabled: no

Layout/AlignHash:
  EnforcedLastArgumentHashStyle: ignore_implicit

Layout/CaseIndentation:
  IndentOneStep: true

Metrics/LineLength:
  Max: 120

Metrics/MethodLength:
  Max: 20

Metrics/AbcSize:
  Max: 30

Metrics/ModuleLength:
  Max: 150

Rails/HasAndBelongsToMany:
  Enabled: no
```

**Comment by jonas054** [comment]
Thanks, that's interesting information. But if this setup cannot reproduce the error, there's probably some interference from other files, so you need to remove the other files one by one, until you no longer get the erroneous behavior. That should move us closer to a solution.

**Comment by zor-el** [comment]
Okay, but how can there be interference across files in the case of `rubocop .` but not with `rubocop */`? The same files should be processed in either case, *or not*?

I think the general assumption here is it is unlikely that `rubocop` would report an `UnneededDisable` if the reported cop was actually enabled at the time of the check. Hence the cop must have been accidentally deactivated in a way that *has to be related to the difference of how files are processed with `rubocop .` vs. `rubocop */`.*

Sorry, this is a project with too many files, and removing each of them one by one and re-checking, that's just not in the cards. Especially if the removal order is relevant in which case I'd have to try all permutations. *Perhaps we can narrow it down by logical deduction first with the above angle as a starting point?*

Some info/thoughts:

[disableScope]: https://stackoverflow.com/q/37921807/4305171

1. Any `# rubocop:disable` [only extends to the end of the file it is in][disableScope]. Hence it cannot come from another source file.
1. The only way, then, to disable a cop, is from within a `.rubocop.yml` file either in the project root directory or in a sub-directory. The latter should only have effect in that directory tree, and there is no other `.rubocop.yml` between the source file in question and the project root.
1. Regardless, no `.rubocop.yml` in the project disables `Rails/ApplicationRecord` or `Rails/Delegate`.

    ```sh
    $ grep 'Rails/' -r . --include='.rubocop.yml'
    ./features/.rubocop.yml:Rails/SkipsModelValidations:
    ./spec/.rubocop.yml:Rails/SkipsModelValidations:
    ./spec/.rubocop.yml:Rails/FilePath:
    ./.rubocop.yml:Rails/HasAndBelongsToMany:

**Comment by zor-el** [comment]
Accidentally closed this :(

**Comment by jonas054** [comment]
> Okay, but how can there be interference across files in the case of rubocop . but not with rubocop */? The same files should be processed in either case, or not?

There is a difference. With `rubocop .` you inspect files in the root (current) directory, which you don't with `rubocop */`. Also, the files are not inspected in the same order between these two commands.

> Any # rubocop:disable only extends to the end of the file it is in. Hence it cannot come from another source file.

Unless there's a bug in RuboCop. :)

There's one important question I haven't asked yet. What happens if you remove `# rubocop:disable Rails/ApplicationRecord`? Do you get a report from `Rails/ApplicationRecord` in that file?

**Comment by zor-el** [comment]
> There's one important question I haven't asked yet. What happens if you remove # rubocop:disable Rails/ApplicationRecord? Do you get a report from Rails/ApplicationRecord in that file?

Yes, I do. That's how I know the UnneededDisable offense is a false positive. :)

> There is a difference. With rubocop . you inspect files in the root (current) directory, which you don't with rubocop */

In the root there is only a Rakefile that would be inspected, and I tried `rubocop Rakefile */` as well, still no false positives.

> Also, the files are not inspected in the same order between these two commands.

But shouldn't they be? Obviously, if `rubocop`'s implementation was perfect, the order wouldn't matter, but bug reproducibility might benefit from it. Just a thought. Anyhow, all `# rubocop:disable Rails/ApplicationRecord` occur on the `class ...` line, and this generally worked.

``` sh
$ grep Rails/App -r . --include=*.rb
./app/models/account.rb:class Account < ActiveRecord::Base # rubocop:disable Rails/ApplicationRecord
./spec/_frameworks/enumerize_spec.rb:class Something < ActiveRecord::Base # rubocop:disable Rails/ApplicationRecord
./spec/_frameworks/role_ranking_spec.rb:class Something < ActiveRecord::Base # rubocop:disable Rails/ApplicationRecord
```

>> Any # rubocop:disable only extends to the end of the file it is in. Hence it cannot come from another source file.

> Unless there's a bug in RuboCop. :)

When I replace the single-line magic comment in all the above files and surround the classes with an enable/disable pair, the false positive still occurs. So even if there were a bug in rubocop that would leak disablings into other files, that bug wouldn't manifest when I re-enable the cops explicitly in the same file. Sadly, the leaking magic disable comment hypothesis seems like a dead end, too. :(

**Comment by jonas054** [comment]
I'm sorry, but I've run out of ideas. I see no way of reproducing the problem. We need a minimal example for reproducing, so if you can come up with that I'd be happy to continue the investigation. Until then I think it's better to close.
